File: ch4/Problems/4.24/monte_carlo_pi_estimation.py
import random
import threading


# Shared global variable to count points inside the circle
points_in_circle = 0


def point_generator(num_points):
    global points_in_circle
    for _ in range(num_points):
        x, y = random.random(), random.random()
        # Check if the point is inside the circle (x^2 + y^2 <= 1)
        if x**2 + y**2 <= 1:
            points_in_circle += 1

if __name__ == "__main__":
    num_threads = 4 # Can be changed later
    total_points = 100000000 # Can be changed later


    threads = []
    points_per_thread = total_points // num_threads

    for _ in range(num_threads):
        thread = threading.Thread(target=point_generator, args=(points_per_thread,))
        thread.start()
        threads.append(thread)

    for thread in threads:
        thread.join()

    
    # Calculate the estimated value of π
    pi_estimation = 4.0 * points_in_circle / total_points

    print("Estimated value of n:", pi_estimation)

File: ch4/Problems/4.24/MonteCarloPiEstimation.java
import java.util.Random;

public class MonteCarloPiEstimation {

    // Shared global variable to count points inside the circle
    private static long pointsInCircle = 0;

    public static void main(String[] args) {
        int numThreads = 1; // You can change the number of threads here
        int totalPoints = 1000000000; // You can change the number of random points generated here

        Thread[] threads = new Thread[numThreads];

        for (int i = 0; i < numThreads; i++) {
            threads[i] = new Thread(new PointGenerator(totalPoints / numThreads));
            threads[i].start();
        }

        for (int i = 0; i < numThreads; i++) {
            try {
                threads[i].join();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }

        // Calculate the estimated value of n
        double piEstimation = 4.0 * pointsInCircle / totalPoints;

        System.out.println("Estimated value of n: " + piEstimation);
    }

    // Thread class to generate random points and count points inside the circle
    private static class PointGenerator implements Runnable {
        private int numPoints;

        public PointGenerator(int numPoints) {
            this.numPoints = numPoints;
        }

        @Override
        public void run() {
            Random random = new Random();
            for (int i = 0; i < numPoints; i++) {
                double x = random.nextDouble();
                double y = random.nextDouble();
                // Check if the point is inside the circle (x^2 + y^2 <= 1)
                if (x * x + y * y <= 1) {
                    synchronized(MonteCarloPiEstimation.class) {
                        pointsInCircle++;
                    }
                }
            }
        }
    }
}